File: src/learning_tf_py/learning_tf_py/tf_listener.py
import rclpy
from rclpy.node import Node
from rclpy.time import Time
from tf2_ros.buffer import Buffer
from tf2_ros.transform_listener import TransformListener
from geometry_msgs.msg import TransformStamped

class TfListener(Node):
    def __init__(self):
        super().__init__("tf_listener_py")
        self.get_logger().info("坐标系监听方创建成功!(Python)")
        self.buffer_ = Buffer()
        self.listener_ = TransformListener(self.buffer_, self)
        self.timer_ = self.create_timer(1.0, self.timer_callback)

    def timer_callback(self):
        if self.buffer_.can_transform("camera", "laser", Time()):
            transform: TransformStamped = self.buffer_.lookup_transform("camera", "laser", Time())
            self.get_logger().info("---------转换成功----------")
            self.get_logger().info(
                f"target_frame: {transform.header.frame_id} "
                f"source_frame: {transform.child_frame_id} "
                f"[x: {transform.transform.translation.x}"
                f" y: {transform.transform.translation.y}"
                f" z: {transform.transform.translation.z}]"
            )
        else:
            self.get_logger().warn("转换异常!")

def main():
    rclpy.init()
    listener = TfListener()
    try:
        rclpy.spin(listener)
    except KeyboardInterrupt:
        pass
    finally:
        listener.destroy_node()
        if rclpy.ok():
            rclpy.shutdown()

if __name__ == "__main__":
    main()

File: src/learning_tf_cpp/src/tf_listener.cpp
#include "rclcpp/rclcpp.hpp"
#include "tf2_ros/buffer.hpp"
#include "tf2_ros/transform_listener.hpp"

using namespace std::chrono_literals;

class TfListener : public rclcpp::Node {
public:
    TfListener() : Node("tf_listener_cpp") {
        RCLCPP_INFO(this->get_logger(), "坐标系监听方创建成功!");
        buffer_ = std::make_unique<tf2_ros::Buffer>(this->get_clock());
        listener_ = std::make_shared<tf2_ros::TransformListener>(*buffer_, this);
        timer_ = this->create_wall_timer(1s, std::bind(&TfListener::timer_callback, this));
    }

private:
    void timer_callback() {
        try {
            auto transform = buffer_->lookupTransform("camera", "laser", tf2::TimePointZero);
            RCLCPP_INFO(this->get_logger(), "----------转换完成----------");
            RCLCPP_INFO(
                this->get_logger(), "target_frame: %s, source_frame: %s, [%.2f, %.2f, %.2f]",
                transform.header.frame_id.c_str(),
                transform.child_frame_id.c_str(),
                transform.transform.translation.x,
                transform.transform.translation.y,
                transform.transform.translation.z
            );
        } catch (const tf2::LookupException &e) {
            RCLCPP_WARN(this->get_logger(), "转换异常: %s", e.what());
        }
    }

    std::unique_ptr<tf2_ros::Buffer> buffer_;
    std::shared_ptr<tf2_ros::TransformListener> listener_;
    rclcpp::TimerBase::SharedPtr timer_;
};

int main(int argc, char **argv) {
    rclcpp::init(argc, argv);

    auto listener = std::make_shared<TfListener>();
    rclcpp::spin(listener);

    rclcpp::shutdown();

    return 0;
}